src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -841,80 +956,182 @@ else if (block instanceof BlockEndPortal || block instanceof BlockMobSpawner ||
-                entityHanging.posX += pos.getX();
-                entityHanging.posY += pos.getY();
-                entityHanging.posZ += pos.getZ();
-                //also sets position based on tile
-                entityHanging.setPosition(
-                        entityHanging.getHangingPosition().getX(),
-                        entityHanging.getHangingPosition().getY(),
-                        entityHanging.getHangingPosition().getZ());
+    /**
+     * Set the currentStructure to null.
+     */
+    public void resetCurrentStructure()
+    {
+        currentStructure = null;
+    }
 
-                entityHanging.setWorld(world);
-                entityHanging.dimension = world.provider.getDimension();
+    private Boolean spawnEntity(@NotNull final Structure.StructureBlock currentBlock)
+    {
+        final Template.EntityInfo entityInfo = currentBlock.entity;
+        if (entityInfo != null && job instanceof JobBuilder && ((JobBuilder) job).getStructure() != null)
+        {
+            final Entity entity = getEntityFromEntityInfoOrNull(entityInfo);
 
-                world.spawnEntityInWorld(entityHanging);
-            }
-            else if (entity instanceof EntityMinecart)
+            if (entity != null && !isEntityAtPosition(entity, world))
             {
-                @Nullable final EntityMinecart minecart = (EntityMinecart) entity;
-                minecart.posX += pos.getX();
-                minecart.posY += pos.getY();
-                minecart.posZ += pos.getZ();
+                final List<ItemStack> request = new ArrayList<>();
 
-                minecart.setWorld(world);
-                minecart.dimension = world.provider.getDimension();
+                if (entity instanceof EntityItemFrame)
+                {
+                    final ItemStack stack = ((EntityItemFrame) entity).getDisplayedItem();
+                    if (stack != null)
+                    {
+                        stack.stackSize++;
+                        request.add(stack);
+                    }
+                    request.add(new ItemStack(Items.ITEM_FRAME, 1));
+                }
+                else if (entity instanceof EntityArmorStand)
+                {
+                    request.add(entity.getPickedResult(new RayTraceResult(worker)));
+                    entity.getArmorInventoryList().forEach(request::add);
+                }
+                else
+                {
+                    request.add(entity.getPickedResult(new RayTraceResult(worker)));
+                }
 
-                world.spawnEntityInWorld(minecart);
+                if (!Configurations.builderInfiniteResources)
+                {
+                    for (final ItemStack stack : request)
+                    {
+                        if (checkOrRequestItems(stack))
+                        {
+                            return false;
+                        }
+                    }
+                    for (final ItemStack stack : request)
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Combine this loop with the one that starts on line 1063. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3047)